### PR TITLE
Refactor to use slash commands.

### PR DESCRIPTION
**Pull request by TheCurle**
With Discord making "read messages" a privileged intent in the somewhat near future, it's clear they want us to move to slash commands.

Now that we have chewutils, it's pretty easy to move things over.

I'm trying to keep everything with feature parity, but improving and fixing things where necessary.
Notably, the forge and mute commands, where the first wasn't accepted the parameter it advertised, and the latter was improved with the addition of Options:
image

I also implemented a PaginatedCommand helper for the multiple commands whose contents can exceed the maximum embed length. This will be refactored into all necessary commands as needed.

Tricks and Quotes will require some thought, as i have had some extra features requested that would fit well with this refactor.
I'll explain in a comment.

I'm currently about half way through moving all of the commands over.

### CODE REVIEW

**Comment by sciwhiz12** [comment]
I would like to note that even though the Read Messages intent will become privileged in the future, to my knoweldge it will operate the same as any other privileged intent: only requiring bot verification if the bot reaches 100 servers. So this doesn't affect us as much, given that MMDBot/Kiri is only present in the MMD server.

**Comment by TheCurle** [comment]
I indeed did intend this by pointing out that it'll be Privileged, but again, it's clear that Discord wants all bots to use this, and there's nothing stopping us right now.

**Comment by TheCurle** [comment]
A couple of notes:

This PR closes one of the TODO List items; 

> Replace help command system (which currently forces help info to DMs)

Was already done by my last two PRs.
> OldChannels command: Non-existing channel/category reply ${channel/category}

Incidental fix from this rewrite.

The refactor to move all of the Quote related commands into one large command seems to work pretty well, and i intend to do the same with tricks in the near future.
I'll likely migrate both to the SQLite database at the same time for ease of developability.

A bunch of incidental bugs were fixed along the way, and javadocs on all of the command classes were changed to actually sum up what the commands were and how they could be used.

This was a surprising amount of fun to do, thanks to the MMD team for the opportunity to work on this.
To the future (and a ton more PRs)!

**Comment by TheCurle** [comment]
Since i had some spare time, i took the liberty of ticking another thing off the todo list:

> Add account creation date to join and ban embeds

The Join embed already has this, i added the same logic to the Ban embed.

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdAbout.java [hunk 1]
```diff
@@ -48,20 +56,21 @@ public CmdAbout() {
         help = "Gives info about this bot.";
         category = new Category("Info");
         guildOnly = true;
+        // TODO: REMOVE BEFORE PUSHING
```
Comment (antoinegag) [inline src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdAbout.java]: Don't forget

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdHelp.java [hunk 2]
```diff
@@ -49,41 +57,48 @@
...
             arguments = "[command]";
+            listener = new CmdHelp.HelpListener();
         }
 
+        /**
+         * Returns the instance of our button listener.
+         * Used for handling the pagination buttons.
+         */
+        public static PaginatedCommand.ButtonListener getListener() {
+        return listener;
```
Comment (antoinegag) [inline src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdHelp.java:80]: Formatting is off by a tab, line under too

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java [hunk 3]
```diff
@@ -48,6 +62,11 @@ public CmdFabricVersion() {
         arguments = "<Minecraft Version>";
         aliases = new String[]{"fabricv"};
         guildOnly = true;
+
+        OptionData data = new OptionData(OptionType.STRING, "version", "The version of Minecraft to check for.").setRequired(false);
```
Comment (antoinegag) [inline src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java:66]: Usually I see command line tools will wrap required argument names in bracket to express that they are not required. For example here `"[version]"`. Would that make sense?

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java [hunk 4]
```diff
@@ -56,29 +75,28 @@ public CmdFabricVersion() {
...
             return;
         }
 
-        var mcVersion = event.getArgs().trim();
-        if (mcVersion.isEmpty()) {
-            mcVersion = MinecraftVersionHelper.getLatest();
-        }
+        String minecraft = MinecraftVersionHelper.getLatest();
+        OptionMapping version = event.getOption("version");
+        if (version != null)
```
Comment (antoinegag) [inline src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java:85]: Not my codebase but in general I strongly recommend to not do 

```
if(bool)
   exp();
``` 

without the braces. Just my 2 cents

Comment (antoinegag) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java:85]: I see it in other places in the changes but your call

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdForgeVersion.java [hunk 5]
```diff
@@ -46,54 +58,62 @@ public CmdForgeVersion() {
...
-        final var changelogLink = Utils.makeHyperlink("Changelog", String.format(
-            "https://files.minecraftforge.net/maven/net/minecraftforge/forge/%1$s-%2$s/forge-%1$s-%2$s-changelog.txt",
+        try {
+            final var latestForgeVersion = latest.getForgeVersion();
+            final var latestForge = latestForgeVersion.getLatest();
+            var recommendedForge = latestForgeVersion.getRecommended();
+            if (recommendedForge == null) {
+                recommendedForge = "none";
+            }
+            final var changelogLink = Utils.makeHyperlink("Changelog", String.format(
```
Comment (antoinegag) [inline src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdForgeVersion.java:103]: This should go in a util class imo, thoughts?

**Review by antoinegag — COMMENTED** [review]
Small nitpicks here and there because I'm nosy :P

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdAbout.java (hunk 1 above)
Comment (TheCurle) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdAbout.java]: Whoops! That shouldn't even be there, let alone forgetting the todo.

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdHelp.java (hunk 2 above)
Comment (TheCurle) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdHelp.java:80]: I hate Intellij's auto formatter sometimes.

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java (hunk 3 above)
Comment (TheCurle) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java:66]: That's a good point, however discord slash commands have strict requirements on what the option names can be.
The argument will show up with a giant "OPTIONAL" next to it anyway:
image

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java (hunk 4 above)
Comment (TheCurle) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java:85]: I'm a proponent of keeping it simple. the indentation makes it clear it's "owned" by the above expression, and so far nobody has complained in over a decade of teaching :P

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdForgeVersion.java (hunk 5 above)
Comment (TheCurle) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdForgeVersion.java:103]: It should, along with a bunch of other things. I may include that in another future PR.

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java (hunk 4 above)
Comment (antoinegag) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdFabricVersion.java:85]: Well I'm complaining 😛 but all good, I see your reasoning

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdHelp.java (hunk 2 above)
Comment (antoinegag) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/bot/info/CmdHelp.java:80]: Me too, me too

File: src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdForgeVersion.java (hunk 5 above)
Comment (antoinegag) [reply src/main/java/com/mcmoddev/mmdbot/modules/commands/general/info/CmdForgeVersion.java:103]: Alright!

**Pull request merged**